source/http_stream.c

```diff
@@ -14,15 +14,437 @@
+        aws_uv_context_enqueue(binding->uv_context, s_on_response_dispatch, binding);
+    }
+    return AWS_OP_SUCCESS;
+}
+
+struct on_body_args {
+    struct http_stream_binding *binding;
+    struct aws_byte_buf chunk;
+};
+
+int s_on_body_params(napi_env env, napi_value *params, size_t *num_params, void *user_data) {
+    struct on_body_args *args = user_data;
+
+    if (napi_get_reference_value(env, args->binding->node_external, &params[0]) ||
+        napi_create_external_arraybuffer(env, args->chunk.buffer, args->chunk.len, NULL, NULL, &params[1])) {
+        return AWS_OP_ERR;
+    }
+
+    *num_params = 2;
+    return AWS_OP_SUCCESS;
+}
+
+void s_on_body_dispatch(void *user_data) {
+    struct on_body_args *args = user_data;
+    struct aws_allocator *allocator = aws_default_allocator();
+    aws_napi_callback_dispatch(&args->binding->on_body, args);
+    aws_byte_buf_clean_up(&args->chunk);
+    aws_mem_release(allocator, args);
+}
+
+int s_on_response_body(struct aws_http_stream *stream, const struct aws_byte_cursor *data, void *user_data) {
+    (void)stream;
+    struct http_stream_binding *binding = user_data;
+    if (AWS_UNLIKELY(!binding->on_body.callback)) {
+        return AWS_OP_SUCCESS;
+    }
+
+    struct aws_allocator *allocator = aws_default_allocator();
+    struct on_body_args *args = aws_mem_calloc(allocator, 1, sizeof(struct on_body_args));
+    AWS_FATAL_ASSERT(args);
+
+    args->binding = binding;
+    if (aws_byte_buf_init_copy_from_cursor(&args->chunk, allocator, *data)) {
+        AWS_FATAL_ASSERT(args->chunk.buffer);
+    }
+
+    aws_uv_context_enqueue(binding->uv_context, s_on_body_dispatch, args);
+
+    return AWS_OP_SUCCESS;
+}
+
+struct on_complete_args {
+    struct http_stream_binding *binding;
+    int error_code;
+};
+
+int s_on_complete_params(napi_env env, napi_value *params, size_t *num_params, void *user_data) {
+    struct on_complete_args *args = user_data;
+
+    if (napi_create_int32(env, args->error_code, &params[0])) {
+        return AWS_OP_ERR;
+    }
+
+    *num_params = 1;
+    return AWS_OP_SUCCESS;
 }
 
-napi_value http_stream_close(napi_env env, napi_callback_info info) {
+void s_on_complete_dispatch(void *user_data) {
+    struct on_complete_args *args = user_data;
+    aws_napi_callback_dispatch(&args->binding->on_complete, args);
+    aws_mem_release(aws_default_allocator(), args);
+}
+
+void s_on_complete(struct aws_http_stream *stream, int error_code, void *user_data) {
+    (void)stream;
+    struct http_stream_binding *binding = user_data;
+    struct aws_allocator *allocator = aws_default_allocator();
+    struct on_complete_args *args = aws_mem_calloc(allocator, 1, sizeof(struct on_complete_args));
+    AWS_FATAL_ASSERT(args);
+    args->binding = binding;
+    args->error_code = error_code;
+    aws_uv_context_enqueue(binding->uv_context, s_on_complete_dispatch, args);
+}
+
+void s_http_stream_binding_finalize(napi_env env, void *finalize_data, void *finalize_hint) {
```

Comment:
I think all of these s_ functions can be static, can you do a pass on that please?